Progress Report June 2022 (#324)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-jun-2022/index.md

+++
date = "2022-07-08T12:00:00-03:00"
title = "Progress Report June 2022"
author = "GoldenX86"
coauthor = "CaptV0rt3x"
forum = 0
+++

Dear yuz-ers, we had fantastic progress during June! Driver bugs are being squashed, there are kernel and CPU improvements, UI and input changes, and more!

<!--more--> 

## PSA for AMD Radeon users (and NVIDIA tags along)

Let’s begin with a driver bug we [mentioned last month](https://yuzu-emu.org/entry/yuzu-progress-report-may-2022/#graphical-changes-driver-issues-and-the-nostalgia-bliss-that-is-the-good-old-64).
The Vulkan extension `VK_KHR_push_descriptor` finally arrived for AMD hardware with driver version 22.5.2 for Windows, but it wasn’t stable. Radeon users would tell you that any game would crash in Vulkan after updating.
To mitigate this, [toastUnlimited](https://github.com/lat9nq) implemented an extension block for the specific Vulkan driver version AMD reports for 22.5.2 (and its equivalent Linux AMDVLK package), 2.0.226.

Skip forward a month and the new 22.6.1 driver is released with `VK_KHR_push_descriptor` fixed!

> But, there is a but, the new driver reports the same Vulkan version, 2.0.226, which forces our devs into a dilemma.

Reviewer: ```suggestion
But, and there is a but, the new driver reports the same Vulkan version, 2.0.226, which forces our devs into a dilemma.
```

The source of the pain was not having any way to use a debugger with the emulated games.

Originally, yuzu inherited a `GDB-compatible debugger interface` from [Citra](https://citra-emu.org), but it was lacking many important features.
And even that had to be deprecated during [Project Prometheus](https://yuzu-emu.org/entry/yuzu-prometheus/) (multicore emulation) because of its inherent shortcomings.

 * It only worked with single core mode
 * It was _slow_ - it could sometimes take 30+ minutes to boot a game, particularly if you had any logging scripts
 * It had some significant code quality issues

After being removed during the Prometheus rewrite, yuzu did not have ***any*** debugger interface for a long time.

### Wait, what is GDB again?

> <p style="color:cyan;">The GNU Debugger (GDB) is a portable debugger that runs on many Unix-like systems and works for many programming languages, including Ada, C, C++, Objective-C, Free Pascal, Fortran, Go, and partially others.</p>
>
> -- [*Wikipedia*](https://en.wikipedia.org/wiki/GNU_Debugger)

With GDB, you can:

>  * step through code on an instruction-by-instruction basis
>  * modify memory and registers on-the-fly
>  * even completely replace sections of running code dynamically

Reviewer: ```suggestion
 * Step through code on an instruction-by-instruction basis
 * Modify memory and registers on-the-fly
 * Even completely replace sections of running code dynamically
```

{{< gh-hovercard "8394" "The new debugger interface" >}} overcomes this by performing debug stepping on threads instead of stepping the emulated CPU core. In yuzu's context, when a thread is stepped, the debugger will ask the thread to step, then the Dynarmic interface will detect this condition and tell Dynarmic to step it, and when the thread has been scheduled again, it will mark that the thread stepped and notify the debugger again.

{{< imgs
    "./gdb2.png| Super Mario Odyssey, in gibberish form"
  >}}

### What are the benefits?

Apart from this, we have a few more notable quality-of-life (QoL) additions.
The debugger interface is now thread-stable, edge cases in stepping and pausing are now handled, and it has tons of useful debugging features, like:

- Support for both 32-bit and 64-bit code
- Ability to modify any memory and registers at any time
- Readout of guest thread names
- Support for unlimited numbers of instruction breakpoints
- Support for up to 4 memory watchpoints

## UI changes

When talking about user interface and experience, you can always count on [Docteh](https://github.com/Docteh).

> [In a repeat of what Morph fixed back in february](https://yuzu-emu.org/entry/yuzu-progress-report-feb-2022/#general-bugfixes-and-ui-changes), Docteh found out that after a crash, the yuzu main window may reopen in some kind of borderless fullscreen… *thing*.

Reviewer: ```suggestion
[In a repeat of what Morph fixed back in February](https://yuzu-emu.org/entry/yuzu-progress-report-feb-2022/#general-bugfixes-and-ui-changes), Docteh found out that after a crash, the yuzu main window may reopen in some kind of borderless fullscreen… *thing*.
```

The culprit was {{< gh-hovercard "8400" "the `UILayout\geometry` value in yuzu’s qt-config.ini file." >}} 
A slap in the face and the issue should be gone for good. Ouch.

> With the intention of helping new users adapt to yuzu, Docteh {{< gh-hovercard "8405" "renamed the status bar" >}} `DOCK` text (which used to only change colour to reflect its status,), to `DOCKED/HANDHELD`.

Reviewer: ```suggestion
With the intention of helping new users adapt to yuzu, Docteh {{< gh-hovercard "8405" "renamed the status bar" >}} `DOCK` text (which used to only change colour to reflect its status) to `DOCKED/HANDHELD`.
```

It was slow, but safe.
`Fire Emblem: Three Houses` would get into a GPU thread race condition with the new method.
{{< gh-hovercard "8483" "Telling the kernel to wait for all threads to stop on pause" >}} avoids the crash.

If, for some reason, yuzu would jump to an invalid address, emulation would hang and the log would get spammed with infinite amounts of `Unmapped Reads`.
Fixing this required work on both Dynarmic and yuzu, resulting in {{< gh-hovercard "8490" "stopping ReadCode callbacks to unmapped addresses." >}}

[exlaunch](https://github.com/shadowninja108/exlaunch) is a framework for injecting C or C++ code into Switch applications and modules.
exlaunch can work on unpatched units, allowing developers to “go to town” with it.
yuzu didn’t support it, but [comex](https://github.com/comex) {{< gh-hovercard "8504" "implemented the required functionality" >}} to have it up and running. 
Thank you!

Newcomer [DCNick3](https://github.com/DCNick3) joins the fray!
For their first brawl, they {{< gh-hovercard "8473" "implemented the `ExitProcess` SVC," >}} which allows homebrew apps to gracefully exit on close.

> ## Issues with third party antiviruses

Reviewer: ```suggestion
## Issues with third-party antiviruses
```

Users have recently reported crashes starting with Mainline version 1075 and newer.
The cause seems to be third-party antiviruses, more specifically ESET/ NOD32. 

> a [HIPS](https://help.eset.com/ees/8/en-US/idh_hips_main.html) false positive is issued, sandboxing yuzu and blocking its access to the system page file.

Reviewer: ```suggestion
A [HIPS](https://help.eset.com/ees/8/en-US/idh_hips_main.html) false positive is issued, sandboxing yuzu and blocking its access to the system page file.
```

{{< gh-hovercard "8394" "The new debugger interface" >}} overcomes this by performing debug stepping on threads instead of stepping the emulated CPU core. In yuzu's context, when a thread is stepped, the debugger will ask the thread to step, then the Dynarmic interface will detect this condition and tell Dynarmic to step it, and when the thread has been scheduled again, it will mark that the thread stepped and notify the debugger again.

{{< imgs
    "./gdb2.png| Super Mario Odyssey, in gibberish form"
  >}}

### What are the benefits?

Apart from this, we have a few more notable quality-of-life (QoL) additions.
The debugger interface is now thread-stable, edge cases in stepping and pausing are now handled, and it has tons of useful debugging features, like:

- Support for both 32-bit and 64-bit code
- Ability to modify any memory and registers at any time
- Readout of guest thread names
- Support for unlimited numbers of instruction breakpoints
- Support for up to 4 memory watchpoints

## UI changes

When talking about user interface and experience, you can always count on [Docteh](https://github.com/Docteh).

> [In reminiscence of what we reported back in february](https://yuzu-emu.org/entry/yuzu-progress-report-feb-2022/#general-bugfixes-and-ui-changes), Docteh found out that after a crash, the yuzu main window may reopen in some kind of borderless fullscreen… *thing*.

Reviewer: ```suggestion
[In a repeat of what Morph fixed back in february](https://yuzu-emu.org/entry/yuzu-progress-report-feb-2022/#general-bugfixes-and-ui-changes), Docteh found out that after a crash, the yuzu main window may reopen in some kind of borderless fullscreen… *thing*.
```

The culprit was {{< gh-hovercard "8400" "the `UILayout\geometry` value in yuzu’s qt-config.ini file." >}} 
A slap in the face and the issue should be gone for good. Ouch.

With the intention of helping new users adapt to yuzu, Docteh {{< gh-hovercard "8405" "renamed the status bar" >}} `DOCK` text (which used to only change colour to reflect its status,), to `DOCKED/HANDHELD`.
Now the current emulated status is clearer, and users won’t confuse it when using dark or light themes. 
Making things easier to understand must never be underestimated. Car makers should try it someday.

{{< single-title-imgs
    "The devil is in the details"
    "./dockbug.png"
    "./dockfix.png"
    >}}

> Translation shenanigans always manage to slip. 

Reviewer: ```suggestion
Translation bugs always manage to slip by. 
```

Additionally, some classes affecting the Web Applet were also deprecated, so {{< gh-hovercard "8477" "some tweaks were in order to ensure future compatibility." >}}
Hopefully Qt6 will mean the return of the Web Applet by default?

Once we’re ready to migrate, this should provide better dynamic DPI scaling, allowing 4K display users to finally understand what is going on in the control settings window, for example.

## Input improvements

Input is [german77’s](https://github.com/german77) speciality, a “diamond in the rough” that is being polished one PR at a time, for eternity….

Continuing the work with `Ring Fit Adventure`, german77 {{< gh-hovercard "8487" "stubbed the `PerformSystemButtonPressingIfInFocus` service" >}}, solving an SVC (Supervisor Call) crash which occurred when pressing ZL or ZR.

With the official Switch update for the firmware version 13.2.0, Nintendo implemented a new `GetVibrationDeviceInfo`. 
While german77 worked on implementing those changes, one game in particular refused to work, `de Blob`.
When this game sends a controller disconnect signal, it uses a `-1` value, which is invalid as only unsigned values are accepted on the Switch.

> Maybe it’s an emulation issue somewhere, or this game just loves to do this and the Switch just accepts invalid values, so our only option is to {{< gh-hovercard "8492" "replicate this peculiar behaviour." >}}
> The end result is allowing `de Blob` to get in-game.

Reviewer: ```suggestion
Maybe it’s an emulation issue somewhere, or this game just loves to do this and the Switch just accepts invalid values. Regardless, our solution is to {{< gh-hovercard "8492" "replicate this peculiar behaviour." >}}
The end result is `de Blob` now gets in-game!
```

{{< imgs
    "./deblob2.png| de Blob 2"
  >}}

## Kernel and CPU changes

Possibly the most silent part of yuzu’s code, but also the most critical.
Kernel emulation is the engine block that keeps all parts working together in harmony, so you can expect that changing even a small part of it can have ripple effects anywhere.

> One must tread carefully, heh, thread carefully.
> I’m not sorry.

Reviewer: ```suggestion
One must tread carefully, heh, thread carefully.
Sorry, not sorry.
```

> Ehem, byte[] has been particularly busy this month in this delicate area, screwdriver in hand and not fearing anything. 

Reviewer: ```suggestion
Anyway, byte[] has been particularly busy this month in this delicate area, screwdriver in hand and not fearing anything. 
```

Clean pause and resume is always a blessing when you have to leave the PC to do something else.

While working on {{< gh-hovercard "8388" "simplifying guest pauses" >}} for single core and multicore emulation, byte[] discovered that if asynchronous GPU emulation and multicore CPU emulation were disabled (something we strongly recommend against, but it’s a valid option for CPU thread starved users, or FX users), a race condition would happen when initializing the CPU and GPU threads.
Several single threads still are multiple threads.
Hammer in hand, byte[] {{< gh-hovercard "8476" "implemented some barriers" >}} to fix this specific crash.

Pause is the word of the week, and this time, it could make a specific game crash.
The old StallCPU behaviour would wait for all thread execution to stop. 
It was slow, but safe.
`Fire Emblem: Three Houses` would get into a GPU thread race condition with the new method.
{{< gh-hovercard "8483" "Telling the kernel to wait for all threads to stop on pause" >}} avoids the crash.

> If for some reason, yuzu would jump to an invalid address, emulation would hang and the log would get spammed with infinite amounts of `Unmapped Reads`.
> Fixing this required both working on Dynarmic and on yuzu, by {{< gh-hovercard "8490" "stopping ReadCode callbacks to unmapped addresses." >}}

Reviewer: ```suggestion
If, for some reason, yuzu would jump to an invalid address, emulation would hang and the log would get spammed with infinite amounts of `Unmapped Reads`.
Fixing this required work on both Dynarmic and yuzu, resulting in {{< gh-hovercard "8490" "stopping ReadCode callbacks to unmapped addresses." >}}
```

[exlaunch](https://github.com/shadowninja108/exlaunch) is a framework for injecting C or C++ code into Switch applications and modules.
exlaunch can work on unpatched units, allowing developers to “go to town” with it.
yuzu didn’t support it, but [comex](https://github.com/comex) {{< gh-hovercard "8504" "implemented the required functionality" >}} to have it up and running. 
Thank you!

Newcomer [DCNick3](https://github.com/DCNick3) joins the fray!

> As their first brawl, they {{< gh-hovercard "8473" "implemented the `ExitProcess` SVC," >}} which allows homebrew apps to gracefully exit on close.

Reviewer: ```suggestion
For their first brawl, they {{< gh-hovercard "8473" "implemented the `ExitProcess` SVC," >}} which allows homebrew apps to gracefully exit on close.
```

## Issues with third party antiviruses

Users have recently reported crashes starting with Mainline version 1075 and newer.

> The cause seems to be third party antiviruses, more specifically ESET/ NOD32. 

Reviewer: ```suggestion
The cause seems to be third-party antiviruses, more specifically ESET/ NOD32. 
```

- The user can disable fastmem from yuzu’s settings, the setting is in `Emulation > Configure… > General > Debug`, from there, enable the option labeled as `Enable CPU Debugging` at the bottom, and from the CPU tab, disable both `Enable Host MMU Emulation` options near the bottom. This will produce a performance loss that can reach up to 30% on some games.
- Testing has shown that adding exceptions for yuzu in ESET doesn’t solve the issue, so the alternative is to uninstall ESET and use Windows Defender instead.

{{< single-title-imgs
    "Here are image examples on how to reach the required options"
    "./fastmem1.png"
    "./fastmem2.png"
    >}}

## Future changes

toastUnlimited has been working on {{< gh-hovercard "8455" "making yuzu compatible to be compiled with" >}} [LLVM Clang](https://clang.llvm.org/) under [MinGW-w64](https://www.mingw-w64.org/).
There are multiple reasons to consider this approach:

- The default compiler we use for Windows builds, MSVC, is currently unstable on its latest 2022 version, forcing us to revert to version 2019, and making yuzu lose some compiler optimizations in the process, losing a bit of performance.

> - GCC 12, the default Linux compiler yuzu uses, has optimizations errors and problems with some warnings, making it unviable at the moment.

Reviewer: ```suggestion
- GCC 12, the default Linux compiler yuzu uses, has optimization errors and problems with some warnings, making it unviable at the moment.
```

- Clang allows for aggressive optimizations that should provide good performance boosts. One example is [Polly](https://polly.llvm.org/).
- Along with GCC, LLVM makes it much easier to produce code optimized for the SSE4.2 instruction set. That’s right Core 2 Duo users, you’re next in line for the chopping block.

The main reason we haven’t switched to this new system by default is `Project Gaia`, or, well, the lack of Gaias currently. 
Some of its changes are mandatory to get Clang builds up and running on Windows. 
While this pull request is completed, its full implementation will be on hold until Gaia is out, which isn’t far away now.

> Get a kettle, boil some [wotah](https://www.youtube.com/watch?v=XE6DT9y7L-w) and make yourself a cuppa tea, because `Project London` has bloody started.

Reviewer: ```suggestion
Get a kettle, boil some [wotah](https://www.youtube.com/watch?v=XE6DT9y7L-w), and make yourself a cuppa tea, because `Project London` has bloody began.
```
I like the alliteration :P
